pageserver: move things around to prepare for decoding logic

VladLazar opened the pull request:
## Problem

We wish to have high level WAL decoding logic in `wal_decoder::decoder` module. 

## Summary of Changes

For this we need the `Value` and `NeonWalRecord` types accessible there, so:
1. Move `Value` and `NeonWalRecord` to `pageserver::value` and `pageserver::record` respectively.
2. Get rid of `pageserver::repository` (follow up from (1))
3. Move PG specific WAL record types to `postgres_ffi::walrecord`. In theory they could live in `wal_decoder`, but it would create a circular dependency between `wal_decoder` and `postgres_ffi`. Long term it makes sense for those types to be PG version specific, so that will work out nicely.
4. Move higher level WAL record types (to be ingested by pageserver) into `wal_decoder::models`

Related: https://github.com/neondatabase/neon/issues/9335
Epic: https://github.com/neondatabase/neon/issues/9329

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

problame commented on libs/pageserver_api/src/record.rs:
```diff
@@ -0,0 +1,113 @@
+impl NeonWalRecord {
+    /// Does replaying this WAL record initialize the page from scratch, or does
+    /// it need to be applied over the previous image of the page?
+    pub fn will_init(&self) -> bool {
```
make sure this continues to get inlined

VladLazar replied:
https://github.com/neondatabase/neon/pull/9504/commits/b4aa4dcd30de0929f13cc8476ca3d9667131fa77

problame commented on libs/pageserver_api/src/value.rs:
```diff
@@ -33,17 +29,17 @@ impl Value {
 
 impl ValueBytes {
-    pub(crate) fn will_init(raw: &[u8]) -> Result<bool, InvalidInput> {
+    pub fn will_init(raw: &[u8]) -> Result<bool, InvalidInput> {
```
Likewise, inlining must continue to happen!

VladLazar replied:
https://github.com/neondatabase/neon/pull/9504/commits/b4aa4dcd30de0929f13cc8476ca3d9667131fa77

problame reviewed (approved):
```
///
/// Note: Parsing some fields is missing, because they're not needed.
///
/// This is similar to the xl_xact_parsed_commit and
/// xl_xact_parsed_abort structs in PostgreSQL, but we use the same
/// struct for commits and aborts.
///
#[derive(Debug)]
pub struct XlXactParsedRecord {
```

diff would be smaller if these hadn't been moved within the same file

---

```
libs/wal_decoder/src/decoder.rs
```

is empty and has no module comment. 


---


```
impl Value {
    pub fn is_image(&self) -> bool {
        matches!(self, Value::Image(_))
    }

    pub fn will_init(&self) -> bool {
        match self {
            Value::Image(_) => true,
            Value::WalRecord(rec) => rec.will_init(),
        }
    }
}
```

These might no longer be inlining after the crate split-up. They should, they'r e on th hot path.

---


The various `fn decode` on the `Xl....` types previously had a higher chance of inlining than they do now.

VladLazar commented:
> ```
> ///
> /// Note: Parsing some fields is missing, because they're not needed.
> ///
> /// This is similar to the xl_xact_parsed_commit and
> /// xl_xact_parsed_abort structs in PostgreSQL, but we use the same
> /// struct for commits and aborts.
> ///
> #[derive(Debug)]
> pub struct XlXactParsedRecord {
> ```
> 
> diff would be smaller if these hadn't been moved within the same file

I've moved it back to its original location within the file in https://github.com/neondatabase/neon/pull/9504/commits/8dcd2c7b4fcb8c11abd4d6d2a5417cc41a2edd02. Not sure the resulting diff is better though.

> ```
> libs/wal_decoder/src/decoder.rs
> ```
> 
> is empty and has no module comment.

That's intentional. The next PR in the series fills it in. I added it to signal to the reviewer that decoding logic will go there.
I can remove it and add it to the next PR if you prefer.

> ```
> impl Value {
>     pub fn is_image(&self) -> bool {
>         matches!(self, Value::Image(_))
>     }
> 
>     pub fn will_init(&self) -> bool {
>         match self {
>             Value::Image(_) => true,
>             Value::WalRecord(rec) => rec.will_init(),
>         }
>     }
> }
> ```
> 
> These might no longer be inlining after the crate split-up. They should, they'r e on th hot path.

Good point. Strongly suggested to the compiler to inline in https://github.com/neondatabase/neon/pull/9504/commits/b4aa4dcd30de0929f13cc8476ca3d9667131fa77. I did a quick check without the inline directive with cachegrind and I think they were already inlined in release builds.

> The various `fn decode` on the `Xl....` types previously had a higher chance of inlining than they do now.

True, let me disassemble and check.

VladLazar commented:
> > The various `fn decode` on the `Xl....` types previously had a higher chance of inlining than they do now.
> 
> True, let me disassemble and check.

I checked both the code on main and on this branch and `XlClogTruncate::decode` is not inlined on either for x86 release builds. `Value::is_image` and `Value::will_init` is indeed inlined in both versions.

skyzh commented on pageserver/src/import_datadir.rs:
```diff
@@ -456,6 +455,7 @@ pub async fn import_wal_from_tar(
             if let Some((lsn, recdata)) = waldecoder.poll_decode()? {
                 let mut decoded = DecodedWALRecord::default();
                 decode_wal_record(recdata, &mut decoded, tline.pg_version)?;
+                // let (ephemeral_file_ready_buf, special_records) = decode_wal_record(recdata, tline.pg_version);
```
rm commented code?

VladLazar replied:
oops. Done in https://github.com/neondatabase/neon/pull/9504/commits/c825b2aafa3ee22f21d63cbfef2cdfb89016dddd

skyzh commented on pageserver/src/tenant.rs:
```diff
@@ -4798,25 +4799,30 @@ mod tests {
+    use timeline::DeltaLayerTestDesc;
     use utils::id::TenantId;
 
+    #[cfg(feature = "testing")]
```
did we enable feature = "testing" for all CI jobs and makefiles? I don't think this feature is by default enabled.

VladLazar replied:
It's always enabled for unit tests and CI builds.

skyzh reviewed (approved):
LGTM

VladLazar commented on libs/pageserver_api/src/record.rs:
```diff
@@ -0,0 +1,113 @@
+use serde::{Deserialize, Serialize};
+use utils::bin_ser::DeserializeError;
+
+/// Each update to a page is represented by a NeonWalRecord. It can be a wrapper
+/// around a PostgreSQL WAL record, or a custom neon-specific "record".
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum NeonWalRecord {
```
> It doesn't feel right to have this in pageserver_api. What API is this part of? Most of the stuff in this module are part of the pageserver HTTP, and this is surely not that.
I presume it's for the safekeeper -> pageserver connection. But we don't have others parts of that API in this module. It's just a stream of WAL at the moment. I guess we're evolving that API to something more complicated; what will the API look like when it's finished?

I broadly feel the same way. Here's what my thought process was:

Safekeepers will send interpreted WAL records to pageservers ([InterpretedWalRecord](https://github.com/neondatabase/neon/blob/7c4807e4b985e3fa40533459c61eda06022a3fe3/libs/wal_decoder/src/models.rs#L22-L37) from a [draft of the follow-up](https://github.com/neondatabase/neon/pull/9524)). These currently include `pageserver_api::value::Value` instances, but eventually these will be replaced with a raw buffer which can be ingested directly by writing to the ephemeral file. I needed the `Value` type to live in `pageserver_api` to enable this arrangement, and, transitively, `NeonWalRecord` also had to move.

Looking back at this PR, I only need the `NeonWalRecord::Postgres` variant for the decoder module. All the other variants will only be generated by the pageserver in `walingest`. I could define a new `RawValue` with just two variants `Image` and `Postgres`, and move `Value` and `NeonWalRecord` back into `pageserver`. I think this might work out nicer. Will try.

VladLazar replied:
I ended up moving `NeonWalRecord` and `Value` back into the pageserver crate. I think it ultimately makes sense for them to live there.

VladLazar replied:
And then I realised that I need access to the `Value` type such that the safekeeper will be able to serialize them into a buffer ready for dumping into the ephemeral file. Moved them back ...

VladLazar replied:
Will resolve and merge to unblock the next PR in the series. Let's move to a Slack thread if needed.

VladLazar commented on libs/wal_decoder/src/models.rs:
```diff
@@ -0,0 +1,145 @@
+//! This module houses types which represent decoded PG WAL records
+//! ready for the pageserver to interpret. They are higher level
+//! than their counterparts in [`postgres_ffi::record`].
```
Good suggestion, thanks! Done in https://github.com/neondatabase/neon/pull/9504/commits/9c5ca8dfc785e356592cd4331e64d13150734190

VladLazar commented on pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:
```diff
@@ -343,6 +343,7 @@ pub(super) async fn handle_walreceiver_connection(
                         let mut decoded = DecodedWALRecord::default();
                         decode_wal_record(recdata, &mut decoded, modification.tline.pg_version)?;
 
+                        // TODO: Handle this. Probably flush buf + data modifications early.
```
Slipped through. Removed in https://github.com/neondatabase/neon/pull/9504/commits/b8f3d56723398c1c5170c87dd99ed7e81c869297

VladLazar commented on libs/postgres_ffi/src/record.rs:
```diff
@@ -1,107 +1,144 @@
+//! This module houses types used in decoding of PG WAL
```
Sure: https://github.com/neondatabase/neon/pull/9504/commits/011494cc9e26493fe1a1ee2111b6337383551925

VladLazar commented on pageserver/src/gc_result.rs:
```diff
@@ -0,0 +1,57 @@
+use anyhow::Result;
```
Good call: https://github.com/neondatabase/neon/pull/9504/commits/99a3f1b3ce4cee699d62b999aefc21509d8fd62a
